Relax the constraint to have a dictionary and inverted index when forward index is disabled

somandal opened the pull request:
This PR relaxes the current constraint that a dictionary and inverted index must exist to allow disabling the forward index.  The following changes have been made as part of this PR:

- Relax the constraint in the `TableConfigUtils` and other places such as segment creator and default index handler
- Update the reload code path to handle toggling the dictionary for forward index disabled columns
- Update the `NoOpForwardIndexCreator` to work for both dictionary based and non-dictionary based columns
- Fix a bug in creating the inverted index column for MV default column with forward index disabled
- Add tests

Adding a new index on a column with forward index disabled along with inverted index or dictionary disabled will result in an error now in reload since the inverted index + dictionary are needed to generate a temporary forward index.

The reload code path needs some significant updates in the `ForwardIndexHandler` to support this change and allow toggling the dictionary for forward index disabled columns. The approach taken to support this can be summarized as follows:

- Forward index -> No Forward Index
    - Sorted columns are ignored (no-op) [current code path]
    - Existing column has dictionary
        - Dictionary being disabled
            - ~~Regenerate temporary forward index to be raw based (post updateIndex cleanup will remove it)~~
            - Remove dictionary
            - Update segment metadata to indicate dictionary disabled
            - Remove all dictionary indexes: inverted index and fst (as they need dictionary) and range (throw error if range index exists)
        - Dictionary remains enabled [current code path]
            - Remove forward index (post updateIndex cleanup)
    - Existing column doesn't have dictionary
        - Dictionary remains disabled
            - Remove forward index (post updateIndex cleanup)
        - Dictionary being enabled [current code path]
            - Regenerate temporary forward index to be dictionary based (post updateIndex cleanup will remove it)
            - Update segment metadata to indicate dictionary enabled
            - Remove all dictionary indexes: inverted index and fst (they shouldn't even exist as they need dictionary) and range (needs to be regenerated and is possible to regenerate on this path)
- No Forward Index -> Forward Index
    - Sorted columns are ignored (no-op) [current code path]
    - If either dictionary or inverted index is disabled on the existing column, log a warning and treat this as a no-op
    - Otherwise regenerate the forward index if dictionary enabled or disabled [current code path]
- No Forward Index -> No Forward Index (this is not getting toggled but maybe something else is)
    - Existing column has dictionary
        - Dictionary being disabled
            - Remove the dictionary
            - Update segment metadata to indicate dictionary disabled
            - Remove all dictionary indexes: inverted index and fst (as they need dictionary) and range (throw error if range index exists)
            - ~~Inverted index exists~~
                - ~~Temporary forward index should be regenerated as raw format from dictionary + inverted index~~
                - ~~Update segment metadata to indicate dictionary disabled~~
                - ~~Remove all dictionary indexes: inverted index and fst (as they need dictionary)  and range (needs to be regenerated and is possible to regenerate on this path)~~
            - ~~Inverted Index does not exist~~
                - ~~Cannot regenerate the forward index in raw format: SHOULD FAIL if RANGE INDEX exists since range index needs to be regenerated, otherwise:~~
                - ~~Just remove dictionary~~
                - ~~Update segment metadata to indicate dictionary disabled~~
                - ~~Remove all dictionary indexes: inverted index and fst (as they need dictionary) and range (needs to be regenerated and cannot be regenerate on this path)~~
        - Dictionary remains enabled [current code path]
            - no-op
    - Existing column doesn't have dictionary
        - Dictionary remains disabled
            - no-op
        - Dictionary being enabled
            - Cannot rebuild dictionary if forward index is not available. Throw an error

cc @siddharthteotia @Jackie-Jiang @vvivekiyer

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/NoOpForwardIndexCreator.java:
```diff
@@ -19,12 +19,13 @@
+ * Forward index creator for dictionary-encoded and raw single and multi-value columns with forward index disabled.
  * This is a no-op.
  */
 public class NoOpForwardIndexCreator implements ForwardIndexCreator {
```
Suggest getting rid of this class and handle disabled forward index explicitly instead of still dumping the values into the no-op creator for both readability and performance concern

somandal replied:
Sure Jackie, but let me handle removing this as a separate change rather than mixing this into this one.

somandal replied:
@Jackie-Jiang I've removed this as part of this PR itself

Jackie-Jiang reviewed (commented):
I feel we are over-complicating this problem. When inverted index does not exist, disabling forward index means only allowing using the index for the column (text, FST). No scan is allowed for the column, and there is no way to recover the column values or create other indexes. On the forward index handler side we can simply remove the forward index without worrying about creating other indexes.

somandal commented:
> I feel we are over-complicating this problem. When inverted index does not exist, disabling forward index means only allowing using the index for the column (text, FST). No scan is allowed for the column, and there is no way to recover the column values or create other indexes. On the forward index handler side we can simply remove the forward index without worrying about creating other indexes.

Yes I agree, for inverted index there isn't much to do and this PR doesn't do any special handling for it except relaxing the constraint. The main complexity here comes from handling changes to the dictionary.

We have a scenario where the dictionary can be disabled for a forward index disabled column. For such scenarios we don't handle anything on reload today. Are you saying that even for toggling whether the dictionary is enabled or disabled we should just treat reload as a no-op rather than throwing error? What about indexes like the range index that need to be modified when the column has the dictionary enabled or disabled?

Jackie-Jiang commented:
> We have a scenario where the dictionary can be disabled for a forward index disabled column. For such scenarios we don't handle anything on reload today. Are you saying that even for toggling whether the dictionary is enabled or disabled we should just treat reload as a no-op rather than throwing error? What about indexes like the range index that need to be modified when the column has the dictionary enabled or disabled?

Disable dictionary should be independent of disabling forward index. Currently it is no-op because we want to ensure data can be recovered. With the new change, we should allow no dictionary + no forward index. Note that disable dictionary + enable inverted index is invalid config, and we should reject it during table config validation.
When both forward index and dictionary are disabled, no further index modification is allowed.

somandal commented:
> > We have a scenario where the dictionary can be disabled for a forward index disabled column. For such scenarios we don't handle anything on reload today. Are you saying that even for toggling whether the dictionary is enabled or disabled we should just treat reload as a no-op rather than throwing error? What about indexes like the range index that need to be modified when the column has the dictionary enabled or disabled?
> 
> Disable dictionary should be independent of disabling forward index. Currently it is no-op because we want to ensure data can be recovered. With the new change, we should allow no dictionary + no forward index. Note that disable dictionary + enable inverted index is invalid config, and we should reject it during table config validation. When both forward index and dictionary are disabled, no further index modification is allowed.

Sure that makes sense. So if I'm understanding correctly I should make the following changes:

- When dictionary is disabled for a forward index disabled column we just remove the dictionary, update the segment metadata and remove the related indexes. If range index is present we can throw an error since we shouldn't try to regenerate it under any circumstances.
- If someone tries to enable the dictionary on a column without forward index, again we should avoid any regeneration business and just throw an error on reload.
- Add validations that inverted index and FST are also disabled if dictionary is disabled on a forward index disabled column (if such a compatibility check doesn't already exist for all columns)

Let me know if the above sounds good or not. I'll keep all code paths marked as [current code path] in the list in the main comment.

somandal commented:
@Jackie-Jiang I addressed your comments, can you take a look when you get a chance? Thanks!

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandler.java:
```diff
@@ -93,8 +93,10 @@ public class ForwardIndexHandler extends BaseIndexHandler {
   protected enum Operation {
-    DISABLE_FORWARD_INDEX_FOR_DICT_COLUMN,
-    DISABLE_FORWARD_INDEX_FOR_RAW_COLUMN,
+    DISABLE_FORWARD_INDEX_FOR_DICT_OR_RAW_COLUMN,
```
This part is very complicated. Are we able to simplify the operations allowed?
I feel we can simplify it by making changes in steps:
1. Check if we need to enable/disable dictionary (switch between dictionary encoding and raw forward index)
2. Check if we need to enable/disable the forward index
3. Check if we need to change the raw index compression type

somandal replied:
Let me look at the code and get back to you about the feasibility of doing the above in steps

somandal replied:
@Jackie-Jiang I was able to take your suggestion and simplify the `ENUMS` used here and perform the operations in steps. Can you take a look and let me know what you think? Thanks!

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/DefaultIndexCreatorProvider.java:
```diff
@@ -76,6 +75,10 @@ public ForwardIndexCreator newForwardIndexCreator(IndexCreationContext.Forward c
       throws Exception {
     if (!context.hasDictionary()) {
       // Dictionary disabled columns
+      if (context.forwardIndexDisabled()) {
```
(minor) Suggest moving the check to the caller side to be consistent with other index creators. Basically when invoking this method, forward index should be created

somandal replied:
done

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -608,119 +568,125 @@ public void indexRow(GenericRow row)
-              long[] longs = new long[length];
-              for (int i = 0; i < length; i++) {
-                longs[i] = (Long) values[i];
+          if (forwardIndexCreator != null) {
+            if (textIndexCreator != null && !shouldStoreRawValueForTextIndex(columnName)) {
```
(minor) Merge these 2 if conditions
```suggestion
          if (forwardIndexCreator != null && textIndexCreator != null && !shouldStoreRawValueForTextIndex(columnName)) {
```

somandal replied:
merging these won't quite work, I need this same `forwardIndexCreator != null` check for the next part where I set the values in the forward index, that's why I kept this check outside.

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandler.java:
```diff
@@ -110,86 +109,81 @@ public ForwardIndexHandler(SegmentDirectory segmentDirectory, IndexLoadingConfig
+                false);
+            if (columnMetadata.hasDictionary()) {
+              if (!segmentWriter.hasIndexFor(column, ColumnIndexType.DICTIONARY)) {
+                throw new IOException(String.format("Dictionary should still exist after rebuilding forward index "
```
Not introduced in this PR, but suggest changing it to `IllegalStateException` since it is not really `IOException`. Same for other places where the operation is illegal

somandal replied:
done

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandler.java:
```diff
@@ -93,12 +94,10 @@ public class ForwardIndexHandler extends BaseIndexHandler {
   private final Schema _schema;
 
   protected enum Operation {
```
(optional) Ideally we can remove the `Operation` enum, and update the index in steps. Can be handled separately

somandal replied:
Sure, let me pick this up as a future fix (along with the changes to remove some of the tests that aren't really needed)

Jackie-Jiang reviewed (approved):
LGTM with minor comments.

Thanks for adding the detailed tests, but it will be good if we can merge some similar tests to keep the test smaller and with the same coverage. It should also help reduce the CI run time. Currently the total lines of test for the forward index handler is probably close to 5K lines, and is quite hard to manage. This can be addressed separately

somandal commented:
> LGTM with minor comments.
> 
> Thanks for adding the detailed tests, but it will be good if we can merge some similar tests to keep the test smaller and with the same coverage. It should also help reduce the CI run time. Currently the total lines of test for the forward index handler is probably close to 5K lines, and is quite hard to manage. This can be addressed separately

Thanks for the detailed review @Jackie-Jiang ! I've addressed most of your comments. Great point about the test explosion. Let me spend some time going over the tests in more detail and come up with that list which can be removed. I'll address this in a follow up (along with changes to remove that Operation enum).
